app/validation/validator.py

```diff
@@ -518,9 +518,38 @@ def _validate_routing_rule(self, rule, answer_ids_with_group_id, block_or_group)
         rule = rule.get('goto')
         if 'when' in rule:
             errors.extend(self._validate_when_rule(rule['when'], answer_ids_with_group_id, block_or_group['id']))
+            for when_dict in rule.get('when'):
+                errors.extend(self._validate_routing_when_dict(when_dict, answer_ids_with_group_id))
 
         return errors
 
+    def _validate_routing_when_dict(self, when_dict, answer_ids_with_parent_id):
+        errors = []
+        when_values = when_dict.get('values', [])
+        when_value = when_dict.get('value')
+
+        if when_value:
+            when_values.append(when_value)
+
+        for rule_value in when_values:
+            valid_rule = self.is_rule_value_valid(answer_ids_with_parent_id, rule_value)
+
+            if not valid_rule:
+                errors.append(Validator._error_message(f'Answer option and routing rule values mismatch, '
+                                                       f'missing answer value: {rule_value}'))
+
+        return errors
+
+    @staticmethod
+    def is_rule_value_valid(answer_ids_with_parent_id, rule_value):
+        if isinstance(rule_value, int) is not True:
```

Comment:
You can save nesting here if you return early.

```suggestion
        if isinstance(rule_value, int) is True:
           return True
        for block in answer_ids_with_parent_id:
           ...
           ...
```

Reply:
Additionally, I think have unit tests for these methods added will help.

Reply:
I test it already in test_invalid_routing_answer_mismatch.json, `{
                  "goto": {
                    "block": "country-interstitial-one",
                    "when": [
                      {
                        "id": "country-checkbox-answer",
                        "condition": "contains",
                        "value": 7
                      }
                    ]
                  }
                },`

Reply:
I don't think that's the best example for testing this single method, we should be explicitly adding tests for each bit of behaviour we introduce.

Reply:
That's done, thanks!